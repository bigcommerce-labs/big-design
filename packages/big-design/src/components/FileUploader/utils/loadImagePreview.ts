export const loadImagePreview = (file: File): Promise<string | null> => {
  if (!file.type.startsWith('image/')) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = (event) => {
      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      const imageUrl = event.target?.result as string;

      resolve(imageUrl);
    };

    reader.onerror = (error) => {
      // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
      reject(error);
    };

    reader.readAsDataURL(file);
  });
};

export const getImagesPreview = async (files: File[]) =>
  files.reduce(async (prevPromise, file, idx) => {
    const prevResult = await prevPromise;
    const imageUrl = await loadImagePreview(file);

    return { ...prevResult, [idx]: imageUrl };
  }, Promise.resolve<Record<number, string | null>>({}));
