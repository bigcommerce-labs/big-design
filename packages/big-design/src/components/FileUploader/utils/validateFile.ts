import { ValidatorConfig } from '../types';

export const validateFileFormat = (file: File | DataTransferItem, fileFormats = '') => {
  const allowedFormats = fileFormats.split(',').map((format) => format.trim());
  const fileExtension = file.type.split('/').pop()?.toLowerCase() ?? '';

  return allowedFormats.some((format) => {
    if (format === '*' || format === '.*') {
      return true;
    } else if (format.startsWith('.')) {
      return format === `.${fileExtension}`;
    } else if (
      format.startsWith('audio/') ||
      format.startsWith('video/') ||
      format.startsWith('image/')
    ) {
      const mediaTypePrefix = format.split('/')[0];

      return file.type.startsWith(`${mediaTypePrefix}/`);
    }

    return true;
  });
};

export const validateFiles = (files: File[], validators: ValidatorConfig[]) =>
  files.flatMap((file, idx) =>
    validators
      .filter(({ validator }) => !validator(file))
      .map(({ message, type }) => ({ file, fileIdx: idx, message, type })),
  );
