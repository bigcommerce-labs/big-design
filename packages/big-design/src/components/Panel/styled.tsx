import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { StyleableH2 } from '../Typography/private';

export const StyledH2 = styled(StyleableH2)`
  flex-grow: 1;

  & ~ .bd-button {
    width: auto;
    margin-top: 0;
  }
`;

StyledH2.defaultProps = { theme: defaultTheme };
