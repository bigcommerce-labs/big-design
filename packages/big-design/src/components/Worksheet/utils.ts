import { Cell, WorksheetError, WorksheetItem } from './types';

export const mergeCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  newCells.reduce(
    (accum, newCell) => {
      const index = oldCells.findIndex(
        (oldCell) => oldCell.rowIndex === newCell.rowIndex && oldCell.columnIndex === newCell.columnIndex,
      );

      if (index > -1) {
        accum[index] = newCell;

        return accum;
      }

      return accum.concat(newCell);
    },
    [...oldCells], // Note: returns a new array every time
  );

export const deleteCells = <T extends WorksheetItem>(oldCells: Cell<T>[], newCells: Cell<T>[]) =>
  oldCells.filter(
    (oldCell) =>
      !newCells.find((newCell) => newCell.columnIndex === oldCell.columnIndex && newCell.rowIndex === oldCell.rowIndex),
  );

export const editedRows = <T extends WorksheetItem>(editedCells: Cell<T>[], rows: T[]) =>
  editedCells.reduce<T[]>((accum, { rowIndex }) => {
    const row = rows[rowIndex];

    // Check to see if the row already exists in accum
    if (accum.find((editedRow) => editedRow === row)) {
      return accum;
    } else {
      // Only append new rows
      return [...accum, row];
    }
  }, []);

export const invalidRows = <T extends WorksheetItem>(invalidCells: Cell<T>[], rows: T[]) => {
  const mapObj = new Map();

  // Create Map with each row and append errors per row
  invalidCells.forEach(({ rowIndex, hash }) => {
    const row = rows[rowIndex];
    if (mapObj.has(row)) {
      const errors = mapObj.get(row);
      mapObj.set(row, new Set([...errors, hash]));
    } else {
      mapObj.set(row, new Set([hash]));
    }
  });

  return Array.from(mapObj).map<WorksheetError<T>>(([item, errors]) => ({
    item,
    errors: Array.from(errors),
  }));
};
