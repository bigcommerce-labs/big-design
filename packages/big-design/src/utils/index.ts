export * from './typedMemo';
export * from './uniqueId';
export * from './warning';
