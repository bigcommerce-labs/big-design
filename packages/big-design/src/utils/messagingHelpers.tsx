import { CheckCircleIcon, ErrorIcon, InfoIcon, WarningIcon } from '@bigcommerce/big-design-icons';
import { ThemeInterface } from '@bigcommerce/big-design-theme';
import React, { ComponentPropsWithoutRef } from 'react';
import { css } from 'styled-components';

import { ButtonProps } from '../components/Button';
import { LinkProps } from '../components/Link';
import { MarginProps } from '../mixins/margins';

export interface SharedMessagingProps extends ComponentPropsWithoutRef<'div'> {
  actions?: MessageAction[];
  header?: string;
  messages: MessageItem[];
  type?: MessagingType;
  onClose?(): void;
}

export type MessagingType = 'success' | 'error' | 'warning' | 'info';

export interface MessageItem {
  text: string;
  link?: MessageLinkItem;
}

export type MessageLinkItem = Pick<LinkProps, 'external' | 'href' | 'target'> & {
  text: string;
};

export interface MessageAction
  extends Omit<ButtonProps, 'children' | 'mobileWidth' | 'variant' | keyof MarginProps> {
  text: string;
  variant?: Exclude<ButtonProps['variant'], 'primary'>;
}

export const getMessagingIcon = (type: MessagingType, condensed?: boolean) => {
  const size = condensed ? 'large' : 'xLarge';

  switch (type) {
    case 'success':
      return <CheckCircleIcon color="success" size={size} />;

    case 'error':
      return <ErrorIcon color="danger" size={size} />;

    case 'warning':
      return <WarningIcon color="warning50" size={size} />;

    case 'info':
      return <InfoIcon color="primary60" size={size} />;
  }
};

export const getBorderStyle = (type: MessagingType, theme: ThemeInterface) => css`
  ${type === 'success' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.success};
  `};

  ${type === 'error' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.danger};
  `};

  ${type === 'warning' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.warning50};
  `};

  ${type === 'info' &&
  css`
    border-left: ${theme.spacing.xxSmall} solid ${theme.colors.primary60};
  `};
`;

export const getActionVariant = (variant: ButtonProps['variant']) =>
  variant === 'primary' ? 'secondary' : variant;
