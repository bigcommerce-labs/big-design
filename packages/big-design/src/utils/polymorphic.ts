// Types from https://github.com/kripod/react-polymorphic-types with a few modifications for Big Design.

// Block external access to auxiliary types
export {};

type Merge<T, U> = Omit<T, keyof U> & U;

type PropsWithAs<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = P & {
  as?: T extends keyof JSX.IntrinsicElements ? (T extends S ? T : never) : never;
};

export type PolymorphicPropsWithoutRef<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  T extends keyof JSX.IntrinsicElements
    ? React.PropsWithoutRef<JSX.IntrinsicElements[T]>
    : React.ComponentPropsWithoutRef<T>,
  PropsWithAs<P, T, S>
>;

export type PolymorphicPropsWithRef<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  T extends keyof JSX.IntrinsicElements
    ? React.PropsWithRef<JSX.IntrinsicElements[T]>
    : React.ComponentPropsWithRef<T>,
  PropsWithAs<P, T, S>
>;

type PolymorphicExoticComponent<
  P = Record<string, unknown>,
  T extends React.ElementType = React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  React.ExoticComponent<P & { [key: string]: unknown }>,
  /**
   * **NOTE**: Exotic components are not callable.
   */
  <InstanceT extends React.ElementType = T>(
    props: PolymorphicPropsWithRef<P, InstanceT, S>,
  ) => React.ReactElement | null
>;

export type PolymorphicForwardRefExoticComponent<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<
  React.ForwardRefExoticComponent<P & { [key: string]: unknown }>,
  PolymorphicExoticComponent<P, T, S>
>;

export type PolymorphicMemoExoticComponent<
  P,
  T extends React.ElementType,
  S extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements,
> = Merge<React.MemoExoticComponent<React.ComponentType<any>>, PolymorphicExoticComponent<P, T, S>>;
