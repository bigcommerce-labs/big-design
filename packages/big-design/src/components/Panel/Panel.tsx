import React, { ComponentPropsWithoutRef, forwardRef, isValidElement, memo, Ref } from 'react';

import { MarginProps } from '../../mixins';
import { excludePaddingProps } from '../../mixins/paddings/paddings';
import { warning } from '../../utils';
import { Badge, BadgeProps } from '../Badge/Badge';
import { Button, ButtonProps } from '../Button';
import { Flex } from '../Flex';
import { Text } from '../Typography';

import { StyledH2, StyledPanel } from './styled';

interface PrivateProps {
  forwardedRef: Ref<HTMLDivElement>;
}

export interface PanelAction extends Omit<ButtonProps, 'children'> {
  text?: string;
}

export interface PanelProps extends ComponentPropsWithoutRef<'div'>, MarginProps {
  children?: React.ReactNode;
  description?: React.ReactNode;
  header?: string;
  headerId?: string;
  action?: PanelAction;
  badge?: BadgeProps;
}

export const RawPanel: React.FC<PanelProps & PrivateProps> = memo(({ forwardedRef, ...props }) => {
  const filteredProps = excludePaddingProps(props);
  const { action, children, description, header, headerId, badge, ...rest } = filteredProps;

  const renderHeader = () => {
    if (!header && !action) {
      return null;
    }

    if (typeof header !== 'string') {
      return null;
    }

    return (
      <Flex flexDirection="row">
        {Boolean(header) && (
          <StyledH2 id={headerId} marginBottom={description ? 'xxSmall' : 'medium'}>
            {header}
            {badge && <Badge marginLeft="xSmall" {...badge} />}
          </StyledH2>
        )}
        {action && <Button {...action}>{action.text}</Button>}
      </Flex>
    );
  };

  const renderDescription = () => {
    if (!description) {
      return null;
    }

    if (typeof description === 'string') {
      return <Text color="secondary60">{description}</Text>;
    }

    if (isValidElement(description)) {
      return description;
    }

    warning('description must be either a string or a ReactNode.');
  };

  return (
    <StyledPanel
      marginBottom="medium"
      {...rest}
      backgroundColor="white"
      padding={{ mobile: 'medium', tablet: 'xLarge' }}
      ref={forwardedRef}
      shadow="raised"
    >
      {renderHeader()}
      {renderDescription()}
      {children}
    </StyledPanel>
  );
});

export const Panel = forwardRef<HTMLDivElement, PanelProps>(
  ({ className, style, ...props }, ref) => <RawPanel {...props} forwardedRef={ref} />,
);

Panel.displayName = 'Panel';
