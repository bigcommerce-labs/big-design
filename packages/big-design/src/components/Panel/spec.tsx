import { theme } from '@bigcommerce/big-design-theme';
import React, { createRef } from 'react';
import 'jest-styled-components';

import { fireEvent, render } from '@test/utils';

import { Panel } from './Panel';

test('render panel', () => {
  const { container } = render(
    <Panel>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('does not forward styles', () => {
  const { container } = render(
    <Panel className="test" style={{ background: 'red' }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  expect(container.getElementsByClassName('test').length).toBe(0);
  expect(container.firstChild).not.toHaveStyle('background: red');
});

test('renders a header and action', () => {
  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action' }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
      fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit
      labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
    </Panel>,
  );

  const header = getByRole('heading');

  expect(header).toBeInTheDocument();
  expect(header.textContent).toBe('Test Header');

  const actionButton = getByRole('button');

  expect(actionButton).toBeInTheDocument();
  expect(actionButton.textContent).toBe('Test Action');
});

test('action options get forwarded to button', () => {
  const onClick = jest.fn();

  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action', onClick }}>
      Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod
    </Panel>,
  );

  fireEvent.click(getByRole('button'));

  expect(onClick).toHaveBeenCalled();
});

test('forwards data attributes', () => {
  const { getByTestId } = render(
    <Panel header="Test Header" data-testid="panel">
      Dolore proident eiusmod sint est enim laboris
    </Panel>,
  );

  const panel = getByTestId('panel');

  expect(panel).toBeInTheDocument();
});

test('forwards ref', () => {
  const ref = createRef<HTMLDivElement>();

  const { container } = render(<Panel ref={ref} />);
  const panel = container.querySelector('div');

  expect(panel).toBe(ref.current);
});

test('ignores padding props', () => {
  const { getByTestId } = render(
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore - ignoring since paddingRight is not a valid prop
    <Panel header="Test Header" data-testid="panel" paddingRight="xxxLarge">
      Dolore proident eiusmod sint est enim laboris
    </Panel>,
  );

  const panel = getByTestId('panel');

  expect(panel).not.toHaveStyle(`padding-right: ${theme.spacing.xxxLarge}`);
});

test("panel action doesn't go to full width", () => {
  const { getByRole } = render(
    <Panel header="Test Header" action={{ text: 'Test Action' }}>
      Test
    </Panel>,
  );

  const button = getByRole('button');

  expect(button).toHaveStyle('width: auto');
});

test('forwards headerId to heading', () => {
  const { getByText } = render(
    <Panel header="Test Header" headerId="test-header">
      Test
    </Panel>,
  );

  expect(getByText('Test Header')).toHaveAttribute('id', 'test-header');
});
