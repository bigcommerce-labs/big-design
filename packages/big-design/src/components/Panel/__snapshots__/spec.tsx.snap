// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render panel 1`] = `
.c1 {
  margin-bottom: 2rem;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  box-shadow: 0px 1px 6px rgba(49,52,64,0.2);
}

.c0 {
  border-radius: 0;
}

@media (min-width:0px) {
  .c1 {
    padding: 1rem;
  }
}

@media (min-width:720px) {
  .c1 {
    padding: 2rem;
  }
}

@media (min-width:720px) {
  .c0 {
    border-radius: 0.25rem;
    box-shadow: 0px 2px 12px rgba(49,52,64,0.2);
  }
}

<div
  class="c0 c1"
  elevation="raised"
>
  Dolore proident eiusmod sint est enim laboris anim minim quis ut adipisicing consectetur officia ex. Ipsum eiusmod fugiat amet pariatur culpa tempor aliquip tempor nisi. Irure esse deserunt nostrud ipsum id adipisicing enim velit labore. Nulla exercitation laborum laboris Lorem irure sit esse nulla mollit aliquip consectetur velit
</div>
`;
